File: examples/python/ds/api/publisher.py
import ds.api._core as _core
import ds.model.publisher as _model
from ds.model._reference import *
from abc import abstractmethod


class Publisher(_model.Publisher, _core.Entity):

    class Parent(_core.Submanager):

        def get(self):
            return self.api.publishers.one(self._.parent_id) if self._.parent_id else None

        def set(self, x):
            self._.parent_id = PublisherRef.valid(x)
            return self.api.publishers.set((self._, Fields.Parent))

    def __init__(self, parent_id=None, name=None, id=None, api=None):
        super().__init__(parent_id, name, id)
        self.api = api
        self._parent = Publisher.Parent(self)

    @property
    def parent(self):
        return self._parent

    def add(self, x):
        if isinstance(x, _core.Engine):
            return x.publishers.add(self)

    def remove(self):
        self.api.publishers.remove(self)
        self.id = None

    def push(self, x:Fields=None):
        return self.api.publishers.set(self if (x is None) else (self, x))

    def pull(self, x:Fields=None):
        return self.api.publishers.get(self, x)


class PublisherManager(_core.PublisherManager):

    def fetch(self, x: object, id:int=None, debug=False) -> Publisher:
        x_ = str(x)
        p = self.one(x_)
        if p is None:
            p = Publisher(id=id, name=x_, api=self.api)
            self.add(p, debug)
        return p

    def _create(self, x):
        return Publisher(id=x[0], parent_id=x[1], name=x[2], api=self.api)

File: examples/php/ds/api/publisher.php
<?php
namespace books;
require_once __DIR__ . '/../data/publisher.php';

// Foreign: 0

class Publisher extends model\Publisher {
    public $api = null;

    function __construct($parent_id=null, $name=null, $id=null, $api=null) {
        parent::__construct($parent_id, $name, $id);
        $this->api = $api;
    }

    function add($x) {
        if ($x instanceof Engine) {
            return $x->publishers->add($this);
        }
    }

    function remove() {
        $this->api->publishers->remove($this);
        $this->id = null;
    }

    function push(?Fields $x=null) {
        return $this->api->publishers->set(is_null($x) ? $this : $x);
    }

    function pull(?Fields $x=null) {
        return $this->api->publishers->get($x);
    }
}


class PublisherManager extends PublisherData {
    private $api;
    
    function __construct(\quartz\Database $db, $api) {
        parent::__construct($db);
        $this->api = $api;
    }

    function fetch($x, ?int $id=null, ?bool $debug=false): Publisher {
        $x_ = strval($x);
        $p = $this->one($x_);
        if (is_null($p)) {
            $p = new Publisher(id:$id, name:$x_, api:$this->api);
            $this->add($p, $debug);
        }
        return $p;
    }

    protected function create($x) {
        return new Publisher(id:$x[0], parent_id:$x[1], name:$x[2], api:$this->api);
    }
}

?>